restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/CallManager.java

```diff
@@ -589,6 +616,71 @@ private void invite(final Object message) throws IOException, NumberParseExcepti
 
     }
 
+    /**
+     * FIXME: duplicated code make into static function or something more optimized
+     * Replace headers
+     * @param SipServletRequest message
+     * @param Map<String, ArrayList<String> > headers
+     */
+    private void addHeadersToMessage(SipServletRequest message, Map<String, ArrayList<String> > headers) {
+
+        if(headers!=null) {
+            for (Map.Entry<String, ArrayList<String>> entry : headers.entrySet()) {
+                //check if header exists
+                String headerName = entry.getKey();
+
+                StringBuilder sb = new StringBuilder();
+                if(entry.getValue() instanceof ArrayList){
+                    for(String pair : entry.getValue()){
+                        sb.append(";").append(pair);
+                    }
+                }
+                if(logger.isDebugEnabled()) {
+                    logger.debug("headerName="+headerName+" headerVal="+message.getHeader(headerName)+" concatValue="+sb.toString());
+                }
+                if(!headerName.equalsIgnoreCase("Request-URI")){
+                    try {
+                        String headerVal = message.getHeader(headerName);
+                        if(headerVal!=null && !headerVal.isEmpty()) {
+                            message.setHeader(headerName , headerVal+sb.toString());
```

Comment:
Again those headers should be validated at API level when we configure the extension and not on a per request basis. If we want validation of headers (which we do only if we ever expose the API for this publicly), then this should be done at API level